mlflow/dspy/callback.py

```diff
@@ -140,35 +144,39 @@ def _start_span(
             else:
                 span = self._client.start_trace(**common_params)
 
-        self._call_id_to_span[call_id] = span
+        token = set_span_in_context(span)
+        self._call_id_to_span[call_id] = SpanWithToken(span, token)
 
     def _end_span(
         self,
         call_id: str,
         outputs: Optional[Any],
         exception: Optional[Exception] = None,
     ):
-        span = self._call_id_to_span.pop(call_id, None)
+        st = self._call_id_to_span.pop(call_id, None)
 
-        if not span:
+        if not st.span:
             _logger.warning(f"Failed to end a span. Span not found for call_id: {call_id}")
             return
 
         status = SpanStatusCode.OK if exception is None else SpanStatusCode.ERROR
 
         if exception:
-            span.add_event(SpanEvent.from_exception(exception))
+            st.span.add_event(SpanEvent.from_exception(exception))
 
         common_params = {
-            "request_id": span.request_id,
+            "request_id": st.span.request_id,
             "outputs": outputs,
             "status": status,
         }
 
-        if span.parent_id:
-            self._client.end_span(span_id=span.span_id, **common_params)
-        else:
-            self._client.end_trace(**common_params)
+        try:
+            if st.span.parent_id:
+                self._client.end_span(span_id=st.span.span_id, **common_params)
+            else:
+                self._client.end_trace(**common_params)
+        finally:
+            detach_span_from_context(st.token)
```

Comment:
Do we want to make this as a context manager?

```suggestion
        with detach_span_from_context(st.token):
            if st.span.parent_id:
                self._client.end_span(span_id=st.span.span_id, **common_params)
            else:
                self._client.end_trace(**common_params)
```

Reply:
Is there any popular library of doing that? Imo having a context manager only for closure is a bit confusing as it is generally used for a pair of operations and when I see a context manager I'd assume sth is started/opened at the beginning of the scope. For this case, `with detach_span_from_context` looks like the span is detached **within** it's scope rather than at the end. 

> Python’s with statement supports the concept of a runtime context defined by a context manager. This is implemented using a pair of methods that allow user-defined classes to define a runtime context that is entered before the statement body is executed and exited when the statement ends:

Reply:
I saw the same pattern here: https://github.com/mlflow/mlflow/pull/13793/files#diff-a87b7250ac1e6118bfc43922fb80b7a20a2c2038b02865923d745c66c681b205R127-R130. I just thought creating a context manger might be useful if we need to do the same thing in more places.

Reply:
Hmm but speaking of DRY, is there difference between repeating try-finally and repeating with statement..?

Reply:
> is there difference between repeating try-finally and repeating with statement..?

Nope :)